feat: add authentication and environment configuration enhancements

joachimdekker opened the pull request:
### Description

This pull request introduces several updates and improvements across authentication, role-based authorization, and environment configuration for the application:

- **Environment Configuration**:  
  Added `ASPNETCORE_ENVIRONMENT` and `DOTNET_ENVIRONMENT` variables to the `launchSettings.json` file to support better local development environment settings.

- **Authentication Middleware and Service**:  
  Integrated an `AuthenticationService` for role-based authentication, along with the `AuthenticationHandler` middleware for request processing. Token handling was enhanced for improved validation and role extraction.

- **Role Claims Fix**:  
  Resolved issues with handling single role claims, allowing both strings and lists to be properly validated.

- **Controller Enhancements**:  
  Refactored `SessionsController`, `CoursesController`, and `UsersController` to inherit from `BaseController`, utilizing shared authentication logic. Added `AssertAuthentication` calls for role-based access control. Introduced `[SwaggerIgnore]` annotations for `FunctionContext` parameters.

- **Role-Based Authorization**:  
  Centralized role definitions and enforced them for `CoursesController` endpoints.

- **Swagger Updates**:  
  Replaced Swagger security scheme ID with "EntraID" to align with the new authentication mechanism. Enhanced Swagger configurations for authorization and anonymous access.

- **OpenID Connect (OIDC) Token Validation**:  
  Implemented OIDC token validation using Microsoft Identity libraries with stricter checks for audience, issuer, and token lifetime.

These updates aim to enhance the security, maintainability, and development experience for the application.

Commit:
Add authentication middleware and enhance Swagger setup

Introduces `AuthenticationHandler` middleware to validate bearer tokens and ensure proper JWT handling. Adds an `AuthenticationService` for token validation. Updates Swagger-specific authorization to allow anonymous access and refines route naming conventions.

joachimdekker committed:
feat(authentication): implement OIDC token validation support

Added OpenID Connect (OIDC) configuration for token validation using Microsoft Identity libraries. This includes fetching signing keys dynamically and validating tokens with stricter parameters such as audience, issuer, and lifetime. Enhanced middleware to store roles and validation results in the request context.

joachimdekker committed:
fix: update Swagger security scheme ID to EntraID

Replaced the security scheme ID from "Bearer" to "EntraID" in Swagger configuration. This change ensures alignment with the updated authentication mechanism.

joachimdekker committed:
feat: integrate authentication middleware and service

Added `AuthenticationService` with role-based authentication logic and registered it in the service container. Integrated `AuthenticationHandler` middleware into the request pipeline to handle authentication. Updated token handling with enhanced validation and proper role extraction.

joachimdekker committed:
feat: add role-based authorization to course operations

Introduced role definitions and a BaseController for centralized authentication handling. Updated CoursesController to enforce role-based access control for all endpoints using the new authorization system.

joachimdekker committed:
feat: add authentication checks and enhance controller inheritance

Refactored `SessionsController`, `CoursesController`, and `UsersController` to inherit from `BaseController` and utilize shared authentication logic. Introduced `AssertAuthentication` calls to enforce role-based access control. Added `[SwaggerIgnore]` annotations for `FunctionContext` parameters where necessary.

joachimdekker committed:
fix: handle single role claims in authentication checks

Previously, the authentication service assumed "roles" claims were always lists, causing issues for single role strings. This update adds handling for single role claims to ensure proper validation against required roles.

joachimdekker committed:
feat: add environment variables to launch settings

Added ASPNETCORE_ENVIRONMENT and DOTNET_ENVIRONMENT variables to the launchSettings.json file for better environment configuration during development. This ensures the application runs with the correct environment settings locally.

joachimdekker committed:
feat: update Swagger security definition for API access

Revised the security scope in Swagger to use "application.fullaccess" instead of "Admin.Admin" for improved clarity and consistency. Removed unnecessary whitespaces and ensured proper formatting across the file.

nicolasantolini commented on AttendanceSystem.Api/AuthenticationService.cs:
```diff
@@ -0,0 +1,30 @@
+        // Get the roles from the claims
+        var roles = rolesClaim as List<object>;
+        if (roles is null)
+        {
```
wow! that's interesting, honestly I have very little experience with this kind of syntax and token authentication. I am curious to know more about it tomorrow!

joachimdekker replied:
Will definitely explain :)

joachimdekker replied:
Oh wait, this can be so much more cleaner...

nicolasantolini commented on AttendanceSystem.Api/Controllers/SessionsController.cs:
```diff
@@ -5,88 +5,84 @@
+    public async Task<IActionResult> GetAllSessions([HttpTrigger(AuthorizationLevel.Anonymous, "get", Route="courses/{courseId}/sessions")] HttpRequest req, FunctionContext ctx, string courseId)
     {
+        // Authorize
+        await AssertAuthentication(ctx, AllowAll);
```
what happens here? I see above that you return a boolean variable that you do not use here, is it because you use the FunctionContext maybe? I don't see why a method should not work if you are not authenticated, basically

joachimdekker replied:
Ah, the AssertAuthentication is a nice way to improve the readability of all the function code. Instead of having

```csharp
if (!authenticationService.IsAuthenticated(ctx, roles)) {
	return NotAuthenticatedResult();
}

// Rest of function here
```

we rely on the `ExceptionToErrorCode` handler which will look for exceptions. Instead of returning a boolean value, the `AssertAuthentication`method (which is in the BaseController) will throw an `NotAuthorizedException`. This will be caught by the handler, and a 401 will be returned.

nicolasantolini replied:
oh yes now I see it! thank you!

joachimdekker committed:
Merge branch 'main' into feat/add-jwt-handling

joachimdekker committed:
refactor: simplify role validation logic in authentication

Replaced verbose role-checking logic with a concise pattern matching approach, improving readability and maintainability. Ensured functionality remains unchanged while reducing redundant code.

joachimdekker committed:
Merge remote-tracking branch 'origin/main' into feat/add-jwt-handling

# Conflicts:
#	AttendanceSystem.Api/Controllers/SessionsController.cs
